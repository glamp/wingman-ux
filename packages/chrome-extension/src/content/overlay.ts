export interface OverlayOptions {
  onSubmit: (note: string, target: any) => void;
  onCancel: () => void;
}

export function createOverlay(options: OverlayOptions) {
  const overlay = document.createElement('div');
  overlay.id = 'wingman-overlay';
  overlay.style.cssText = `
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 2147483647;
    pointer-events: none;
  `;

  const highlighter = document.createElement('div');
  highlighter.id = 'wingman-highlighter';
  highlighter.style.cssText = `
    position: fixed;
    border: 2px solid #0084ff;
    background: rgba(0, 132, 255, 0.1);
    pointer-events: none;
    transition: all 0.1s ease;
  `;

  const notePanel = document.createElement('div');
  notePanel.id = 'wingman-note-panel';
  notePanel.style.cssText = `
    position: fixed;
    bottom: 20px;
    right: 20px;
    background: white;
    border: 1px solid #e0e0e0;
    border-radius: 8px;
    padding: 16px;
    box-shadow: 0 4px 12px rgba(0,0,0,0.15);
    pointer-events: all;
    display: none;
    z-index: 2147483648;
    width: 320px;
  `;

  const textarea = document.createElement('textarea');
  textarea.placeholder = 'Describe the issue...';
  textarea.style.cssText = `
    width: 100%;
    height: 80px;
    border: 1px solid #e0e0e0;
    border-radius: 4px;
    padding: 8px;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
    font-size: 14px;
    resize: vertical;
    margin-bottom: 12px;
  `;

  const buttonContainer = document.createElement('div');
  buttonContainer.style.cssText = `
    display: flex;
    gap: 8px;
    justify-content: flex-end;
  `;

  const cancelButton = document.createElement('button');
  cancelButton.textContent = 'Cancel';
  cancelButton.style.cssText = `
    padding: 6px 16px;
    border: 1px solid #e0e0e0;
    background: white;
    border-radius: 4px;
    font-size: 14px;
    cursor: pointer;
  `;

  const submitButton = document.createElement('button');
  submitButton.textContent = 'Send';
  submitButton.style.cssText = `
    padding: 6px 16px;
    border: none;
    background: #0084ff;
    color: white;
    border-radius: 4px;
    font-size: 14px;
    cursor: pointer;
  `;

  buttonContainer.appendChild(cancelButton);
  buttonContainer.appendChild(submitButton);
  notePanel.appendChild(textarea);
  notePanel.appendChild(buttonContainer);

  overlay.appendChild(highlighter);
  overlay.appendChild(notePanel);
  document.body.appendChild(overlay);

  let selectedElement: HTMLElement | null = null;
  let selectedRect: DOMRect | null = null;
  let mode: 'element' | 'region' = 'element';

  // Element selection mode
  overlay.style.pointerEvents = 'all';
  
  const handleMouseMove = (e: MouseEvent) => {
    if (mode === 'element') {
      const elements = document.elementsFromPoint(e.clientX, e.clientY);
      const element = elements.find(el => 
        el !== overlay && 
        el !== highlighter && 
        el !== notePanel &&
        !overlay.contains(el) &&
        !notePanel.contains(el)
      );
      
      if (element) {
        const rect = element.getBoundingClientRect();
        highlighter.style.left = `${rect.left}px`;
        highlighter.style.top = `${rect.top}px`;
        highlighter.style.width = `${rect.width}px`;
        highlighter.style.height = `${rect.height}px`;
        highlighter.style.display = 'block';
      } else {
        highlighter.style.display = 'none';
      }
    }
  };

  const handleClick = (e: MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    const elements = document.elementsFromPoint(e.clientX, e.clientY);
    const element = elements.find(el => 
      el !== overlay && 
      el !== highlighter && 
      el !== notePanel &&
      !overlay.contains(el) &&
      !notePanel.contains(el)
    );
    
    if (element) {
      selectedElement = element as HTMLElement;
      selectedRect = element.getBoundingClientRect();
      
      // Show note panel
      notePanel.style.display = 'block';
      textarea.focus();
      
      // Stop highlighting
      overlay.removeEventListener('mousemove', handleMouseMove);
      overlay.style.pointerEvents = 'none';
      notePanel.style.pointerEvents = 'all';
    }
  };

  overlay.addEventListener('mousemove', handleMouseMove);
  overlay.addEventListener('click', handleClick);

  cancelButton.addEventListener('click', () => {
    cleanup();
    options.onCancel();
  });

  submitButton.addEventListener('click', () => {
    const note = textarea.value.trim();
    if (!note) {
      textarea.focus();
      return;
    }

    const target = {
      mode: mode as 'element' | 'region',
      rect: selectedRect ? {
        x: selectedRect.left,
        y: selectedRect.top,
        width: selectedRect.width,
        height: selectedRect.height,
      } : null,
      selector: selectedElement ? generateSelector(selectedElement) : undefined,
    };

    cleanup();
    options.onSubmit(note, target);
  });

  // ESC key to cancel
  const handleKeydown = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      cleanup();
      options.onCancel();
    }
  };
  document.addEventListener('keydown', handleKeydown);

  function cleanup() {
    overlay.remove();
    document.removeEventListener('keydown', handleKeydown);
  }

  return { cleanup };
}

function generateSelector(element: HTMLElement): string {
  const path: string[] = [];
  let current: HTMLElement | null = element;

  while (current && current !== document.body) {
    let selector = current.tagName.toLowerCase();
    
    if (current.id) {
      selector = `#${current.id}`;
      path.unshift(selector);
      break;
    }
    
    if (current.className) {
      const classes = Array.from(current.classList)
        .filter(c => !c.startsWith('wingman-'))
        .join('.');
      if (classes) {
        selector += `.${classes}`;
      }
    }
    
    const siblings = Array.from(current.parentNode?.children || []);
    if (siblings.length > 1) {
      const index = siblings.indexOf(current) + 1;
      selector += `:nth-child(${index})`;
    }
    
    path.unshift(selector);
    current = current.parentElement;
  }

  return path.join(' > ');
}