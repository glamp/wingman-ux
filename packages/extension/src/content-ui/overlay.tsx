import React from 'react';
import App from './App';
import { mountReactInShadow, ShadowMountResult } from './utils/shadowMount';
import type { WingmanAnnotation } from '@wingman/shared';

export interface MountOverlayOptions {
  onSubmit: (note: string, target: any, element?: HTMLElement, screenshot?: string) => void;
  onCancel: () => void;
}

export interface MountSuccessNotificationOptions {
  previewUrl?: string;
  annotation?: WingmanAnnotation;
  mode?: 'clipboard' | 'server';
  onClose: () => void;
}

let currentMount: ShadowMountResult | null = null;

/**
 * Mount the React overlay for element selection and note input
 */
export function mountReactOverlay(options: MountOverlayOptions): () => void {
  // Unmount any existing mount
  if (currentMount) {
    currentMount.unmount();
    currentMount = null;
  }

  console.log('[Wingman] Mounting React overlay...');
  
  currentMount = mountReactInShadow({
    hostId: 'wingman-overlay-host',
    component: App,
    props: {
      mode: 'overlay',
      onSubmit: (note: string, target: any, element?: HTMLElement, screenshot?: string) => {
        // Don't unmount here - let the handler do it after screenshot
        // Call the original handler
        options.onSubmit(note, target, element, screenshot);
      },
      onCancel: () => {
        // Clean up the overlay
        if (currentMount) {
          currentMount.unmount();
          currentMount = null;
        }
        // Call the original handler
        options.onCancel();
      },
    },
  });

  // Return cleanup function
  return () => {
    if (currentMount) {
      currentMount.unmount();
      currentMount = null;
    }
  };
}

/**
 * Mount the React success notification
 */
export function mountSuccessNotification(options: MountSuccessNotificationOptions): () => void {
  console.log('[Wingman] Mounting React success notification...');
  
  const mount = mountReactInShadow({
    hostId: 'wingman-success-host',
    component: App,
    props: {
      mode: 'success',
      previewUrl: options.previewUrl,
      annotation: options.annotation,
      notificationMode: options.mode,
      onNotificationClose: () => {
        // Clean up the notification
        mount.unmount();
        // Call the original handler
        options.onClose();
      },
    },
  });

  // Return cleanup function
  return () => {
    mount.unmount();
  };
}

// Export for direct use if needed
export { App, mountReactInShadow };