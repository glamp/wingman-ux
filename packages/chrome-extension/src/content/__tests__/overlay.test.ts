import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { createSuccessNotification } from '../overlay';

describe('Success Notification', () => {
  beforeEach(() => {
    // Reset DOM
    document.body.innerHTML = '';
    
    // Mock clipboard API
    Object.defineProperty(navigator, 'clipboard', {
      value: {
        writeText: vi.fn().mockResolvedValue(undefined)
      },
      writable: true,
      configurable: true
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('createSuccessNotification', () => {
    it('should create notification with correct structure', () => {
      const onClose = vi.fn();
      const previewUrl = 'http://localhost:8787/preview/?id=test-123';
      
      createSuccessNotification({
        previewUrl,
        onClose
      });

      const notification = document.getElementById('wingman-success-notification');
      expect(notification).toBeTruthy();
      
      // Check for key elements
      expect(notification?.textContent).toContain('Feedback submitted successfully!');
      expect(notification?.textContent).toContain('Preview URL:');
      
      // Check URL input exists and has correct value
      const urlInput = notification?.querySelector('input') as HTMLInputElement;
      expect(urlInput).toBeTruthy();
      expect(urlInput.value).toBe(previewUrl);
      
      // Check buttons exist
      expect(notification?.textContent).toContain('Open Preview');
      expect(notification?.textContent).toContain('Copy for Claude Code');
    });

    it('should handle close button click', () => {
      const onClose = vi.fn();
      
      const { close } = createSuccessNotification({
        previewUrl: 'http://localhost:8787/preview/?id=test',
        onClose
      });

      const closeButton = document.querySelector('button') as HTMLButtonElement;
      expect(closeButton).toBeTruthy();
      
      // Click close button
      closeButton.click();
      
      // Wait for animation
      setTimeout(() => {
        expect(onClose).toHaveBeenCalled();
        expect(document.getElementById('wingman-success-notification')).toBeFalsy();
      }, 350);
    });

    it('should copy URL when copy button is clicked', async () => {
      const previewUrl = 'http://localhost:8787/preview/?id=test-copy';
      
      createSuccessNotification({
        previewUrl,
        onClose: vi.fn()
      });

      // Find the copy button (has ðŸ“‹ emoji)
      const buttons = Array.from(document.querySelectorAll('button'));
      const copyButton = buttons.find(btn => btn.textContent === 'ðŸ“‹');
      expect(copyButton).toBeTruthy();
      
      // Click copy button
      await copyButton?.click();
      
      // Check clipboard was called with correct URL
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(previewUrl);
    });

    it('should open preview in new tab when Open Preview is clicked', () => {
      const previewUrl = 'http://localhost:8787/preview/?id=test-open';
      const windowOpenSpy = vi.spyOn(window, 'open').mockImplementation(() => null);
      
      createSuccessNotification({
        previewUrl,
        onClose: vi.fn()
      });

      // Find the Open Preview button
      const buttons = Array.from(document.querySelectorAll('button'));
      const openButton = buttons.find(btn => btn.textContent === 'Open Preview');
      expect(openButton).toBeTruthy();
      
      // Click open button
      openButton?.click();
      
      // Check window.open was called with correct URL
      expect(windowOpenSpy).toHaveBeenCalledWith(previewUrl, '_blank');
      
      windowOpenSpy.mockRestore();
    });

    it('should copy URL when Copy for Claude Code is clicked', async () => {
      const previewUrl = 'http://localhost:8787/preview/?id=test-claude';
      
      createSuccessNotification({
        previewUrl,
        onClose: vi.fn()
      });

      // Find the Claude button
      const buttons = Array.from(document.querySelectorAll('button'));
      const claudeButton = buttons.find(btn => btn.textContent === 'Copy for Claude Code');
      expect(claudeButton).toBeTruthy();
      
      // Click Claude button
      await claudeButton?.click();
      
      // Check clipboard was called
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(previewUrl);
      
      // Check button text changed to indicate success
      expect(claudeButton?.textContent).toBe('âœ“ Copied!');
    });

    it('should auto-dismiss after 10 seconds', () => {
      vi.useFakeTimers();
      const onClose = vi.fn();
      
      createSuccessNotification({
        previewUrl: 'http://localhost:8787/preview/?id=test',
        onClose
      });

      const notification = document.getElementById('wingman-success-notification');
      expect(notification).toBeTruthy();
      
      // Fast-forward 10 seconds
      vi.advanceTimersByTime(10000);
      
      // Wait for close animation
      vi.advanceTimersByTime(300);
      
      expect(onClose).toHaveBeenCalled();
      
      vi.useRealTimers();
    });

    it('should pause auto-dismiss on hover', () => {
      vi.useFakeTimers();
      const onClose = vi.fn();
      
      createSuccessNotification({
        previewUrl: 'http://localhost:8787/preview/?id=test',
        onClose
      });

      const notification = document.getElementById('wingman-success-notification');
      expect(notification).toBeTruthy();
      
      // Fast-forward 5 seconds
      vi.advanceTimersByTime(5000);
      
      // Simulate mouse enter
      const mouseEnter = new MouseEvent('mouseenter');
      notification!.dispatchEvent(mouseEnter);
      
      // Fast-forward another 10 seconds (should not close while hovering)
      vi.advanceTimersByTime(10000);
      
      expect(onClose).not.toHaveBeenCalled();
      
      // Simulate mouse leave
      const mouseLeave = new MouseEvent('mouseleave');
      notification!.dispatchEvent(mouseLeave);
      
      // Fast-forward 10 seconds after mouse leave
      vi.advanceTimersByTime(10000);
      vi.advanceTimersByTime(300); // animation time
      
      expect(onClose).toHaveBeenCalled();
      
      vi.useRealTimers();
    });

    it('should handle clipboard API failure gracefully', async () => {
      // Mock clipboard to reject
      navigator.clipboard.writeText = vi.fn().mockRejectedValue(new Error('Clipboard failed'));
      
      // Mock document.execCommand as fallback
      document.execCommand = vi.fn().mockReturnValue(true);
      
      const previewUrl = 'http://localhost:8787/preview/?id=test-fallback';
      
      createSuccessNotification({
        previewUrl,
        onClose: vi.fn()
      });

      // Find copy button
      const buttons = Array.from(document.querySelectorAll('button'));
      const copyButton = buttons.find(btn => btn.textContent === 'ðŸ“‹');
      
      // Click copy button
      await copyButton?.click();
      
      // Wait for promise to reject and fallback to execute
      await new Promise(resolve => setTimeout(resolve, 0));
      
      // Check fallback was used
      expect(document.execCommand).toHaveBeenCalledWith('copy');
    });

    it('should create valid snapshot of notification HTML', () => {
      createSuccessNotification({
        previewUrl: 'http://localhost:8787/preview/?id=snapshot-test',
        onClose: vi.fn()
      });

      const notification = document.getElementById('wingman-success-notification');
      
      // Create a cleaner snapshot by removing dynamic styles
      const clone = notification?.cloneNode(true) as HTMLElement;
      if (clone) {
        // Remove inline styles for cleaner snapshot
        clone.removeAttribute('style');
        // Remove all button hover handlers
        clone.querySelectorAll('button').forEach(btn => {
          btn.removeAttribute('style');
        });
      }
      
      expect(clone?.outerHTML).toMatchSnapshot();
    });
  });
});