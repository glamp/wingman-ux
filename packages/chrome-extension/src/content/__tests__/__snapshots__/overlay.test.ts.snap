// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Success Notification > createSuccessNotification > should create valid snapshot of notification HTML 1`] = `"<div id="wingman-success-notification"><button>âœ•</button><div style="display: flex; align-items: center; gap: 10px; margin-bottom: 16px;"><div style="font-size: 24px;">âœ…</div><div style="font-size: 16px; font-weight: 600; color: rgb(30, 41, 59);">Feedback submitted successfully!</div></div><div style="background: rgb(248, 250, 252); border: 1px solid rgb(226, 232, 240); border-radius: 8px; padding: 12px; margin-bottom: 12px;"><div style="font-size: 12px; color: rgb(100, 116, 139); margin-bottom: 6px; font-weight: 500;">Preview URL:</div><div style="display: flex; gap: 8px; align-items: center;"><input readonly="" style="flex: 1 1 0%; padding: 8px; border: 1px solid rgb(226, 232, 240); border-radius: 6px; font-size: 12px; font-family: &quot;SF Mono&quot;, Monaco, &quot;Cascadia Code&quot;, monospace; background: white; color: rgb(30, 41, 59);"><button title="Copy URL">ðŸ“‹</button></div></div><div style="display: flex; gap: 8px;"><button>Open Preview</button><button>Copy for Claude Code</button></div></div>"`;
